File: AtCoder/ARC-C/035probC.cpp
#include <bits/stdc++.h>
typedef long long ll;
#define REP(i, n) for(int i = 0; i < (int)(n); i++)
#define FOR(i, a, b) for(int i = (int)(a); i < (int)(b); i++)
using namespace std;

ll INF = 100000000000000;
ll a, b, c;

int main() {
    ll N, M; cin >> N >> M;
    ll dp[N][N];
    REP(i, N) REP(j, N) dp[i][j] = INF;
    REP(i, M){
        cin >> a >> b >> c;
        if (c < dp[a-1][b-1]){
            dp[a-1][b-1] = c;
        }
        if (c < dp[b-1][a-1]){
            dp[b-1][a-1] = c;
        }
    }
    ll K; cin >> K;
    ll XYZ[K][3];
    REP(i, K){
        cin >> XYZ[i][0] >> XYZ[i][1] >> XYZ[i][2];
    }

    REP(k, N) REP(i, N) REP(j, N){
        dp[i][j] = min(dp[i][j], dp[i][k]+dp[k][j]);
    }
    REP(i, N) dp[i][i] = 0;
    ll x, y, z;
    REP(k, K){
        x = XYZ[k][0];
        y = XYZ[k][1];
        z = XYZ[k][2];
        REP(i, N) REP(j, N){
            dp[i][j] = min(dp[i][j], dp[i][x-1]+z+dp[y-1][j]);
            dp[i][j] = min(dp[i][j], dp[i][y-1]+z+dp[x-1][j]);
        }
        ll ans = 0;
        REP(i, N-1) FOR(j, i+1, N){
            ans += dp[i][j];
        }
        cout << ans << endl;
    }

}

File: AtCoder/ARC-C/035probC.py
import sys
input = sys.stdin.readline

INF = 10**14

def main():
    N, M = map(int, input().split())
    dp = [INF for _ in range(N*N)]

    for _ in range(M):
        a, b, c = map(int, input().split())
        dp[(a-1)*N+b-1] = min(c, dp[(a-1)*N+b-1])
        dp[(b-1)*N+a-1] = min(c, dp[(b-1)*N+a-1])
    K = int(input())
    XYZ = [list(map(int, input().split())) for _ in range(K)]

    for k in range(N):
        for i in range(N):
            for j in range(N):
                dp[i*N+j] = min(dp[i*N+j], dp[i*N+k]+dp[k*N+j])

    for i in range(N):
        dp[i*N+i] = 0

    for x, y, z in XYZ:
        for i in range(N):
            for j in range(N):
                dp[i*N+j] = min((dp[i*N+j], dp[i*N+x-1]+z+dp[(y-1)*N+j], dp[i*N+y-1]+z+dp[(x-1)*N+j]))
        s = 0
        for i in range(N-1):
            for j in range(i+1, N):
                s += dp[i*N+j]
        print(s)
    

if __name__ == "__main__":
    main()